### PR TITLE
Insert/remove range with bounds

### PR DESCRIPTION
**Pull request by oliverdding**
Implimentation of insert_range and remove_range functions that accept RangeBounds instead of Range.

This PR fix [#5] and [#113].

Done:

 * [x] Switch inner parameters to RangeBounds
   * [x] insert_range
   * [x] remove_range
 * [x] Fix boundary condition in insert_range/remove_range
 * [x] Add more test cases
 * [x] switching to RangeBounds of container
   * [x] insert_range
   * [x] remove_range

### CODE REVIEW

File: src/bitmap/store.rs [hunk 1]
```diff
@@ -146,28 +166,36 @@ impl Store {
...
-                let start_bit = bit(start as u16) as u32;
-                // end_key is inclusive
-                let end_key = key((end - 1) as u16) as usize;
-                let end_bit = bit(end as u16) as u32;
+                let (start_key, start_bit) = (key(start), bit(start));
+                let (end_key, end_bit) = (key(end), bit(end));
 
                 if start_key == end_key {
-                    let mask = (!0u64 << start_bit) & (!0u64).wrapping_shr(64 - end_bit);
+                    let mask = (!0u64 << start_bit) & (!0u64 >> (63 - end_bit));
```
Comment (Kerollmops) [inline src/bitmap/store.rs:198]: Why do you remove the `wrapping_shr` here?

File: src/bitmap/store.rs [hunk 2]
```diff
@@ -189,8 +217,8 @@ impl Store {
                     *word = 0;
                 }
                 // end key bits
-                removed += (bits[end_key] & (!0u64).wrapping_shr(64 - end_bit)).count_ones();
-                bits[end_key] &= !(!0u64).wrapping_shr(64 - end_bit);
+                removed += (bits[end_key] & (!0u64 >> (63 - end_bit))).count_ones();
+                bits[end_key] &= !(!0u64 >> (63 - end_bit));
```
Comment (Kerollmops) [inline src/bitmap/store.rs:221]: Same here, why do you remove the `wrapping_shr`?

File: src/treemap/util.rs [hunk 3]
```diff
@@ -8,6 +10,31 @@ pub fn join(high: u32, low: u32) -> u64 {
...
+        Bound::Included(&i) => i,
+        Bound::Excluded(&i) => i - 1,
+        Bound::Unbounded => u64::MAX,
+    };
+    if end < start {
+        return (0..=0, true);
+    }
+    (start..=end, false)
+}
+
```
Comment (Kerollmops) [inline src/treemap/util.rs:37]: Can't you make it generic for `u32`s and `u64`s, don't do it if you need to introduce a new crate i.e. `num-trait`, I prefer that you duplicate this function for u64s like you have done instead of adding a new dep just for that :)

Comment (Kerollmops) [reply src/treemap/util.rs:37]: Can you please apply the same changes as the other `convert_range_to_inclusive` function, please?

**Review by Kerollmops — CHANGES_REQUESTED** [review]

**Comment by oliverdding** [comment]
Hi @Kerollmops 

Thanks for your reviewing!

Recently I might have no time to code since I'm on my way to travel 👻. But I would come back ASAP.

File: src/bitmap/store.rs (hunk 2 above)
Comment (oliverdding) [reply src/bitmap/store.rs:221]: Let's talk about the `wrapping_shr` in the upper conversation and I would mark the below resolved.

File: src/bitmap/util.rs [hunk 4]
```diff
@@ -12,6 +14,31 @@ pub fn join(high: u16, low: u16) -> u32 {
...
+where
+    R: RangeBounds<u32>,
+{
+    if let Bound::Excluded(0) = range.end_bound() {
+        return (0..=0, true);
+    }
+    let start: u32 = match range.start_bound() {
+        Bound::Included(&i) => i,
+        Bound::Unbounded => 0,
+        _ => panic!("Should never be called (insert_range start with Excluded)"),
```
[the remark this answers is not in the document]
Comment (oliverdding) [reply src/bitmap/util.rs:28]: Ok, I have no idea about `saturating_sub`, I would give it a look ;)

File: src/treemap/inherent.rs [hunk 5]
```diff
@@ -113,14 +113,14 @@ impl RoaringTreemap {
...
     {
-        let (range, is_empty) = util::convert_range_to_inclusive(range);
-        if is_empty {
+        let (start, end);
+        if let Some(range) = util::convert_range_to_inclusive(range) {
+            start = *range.start();
+            end = *range.end();
+        } else {
             return 0;
         }
```
Comment (Kerollmops) [inline src/treemap/inherent.rs:122]: I would prefer that we use a match here instead, it is easier to read and more rust idiomatic :)
Can you do this at the other places in your changes, please?

```suggestion
        let (start, end) = match util::convert_range_to_inclusive(range) {
            Some(range) => (*range.start(), *range.end()),
            None => return 0,
        };
```

File: src/bitmap/inherent.rs [hunk 6]
```diff
@@ -76,14 +76,14 @@ impl RoaringBitmap {
     where
         R: RangeBounds<u32>,
     {
-        let (range, is_empty) = util::convert_range_to_inclusive(range);
-        if is_empty {
+        let (start, end);
```
Comment (Kerollmops) [inline src/bitmap/inherent.rs:79]: Can you use a match instead, please?

File: src/bitmap/inherent.rs [hunk 7]
```diff
@@ -207,14 +207,14 @@ impl RoaringBitmap {
...
```
Comment (Kerollmops) [inline src/bitmap/inherent.rs:210]: Can you use a match instead, please?

File: src/bitmap/util.rs [hunk 8]
```diff
@@ -12,6 +14,31 @@ pub fn join(high: u16, low: u16) -> u32 {
...
+    };
+    let end: u32 = match range.end_bound() {
+        Bound::Included(&i) => i,
+        Bound::Excluded(&i) => i - 1,
```
[the remark this answers is not in the document]
Comment (oliverdding) [reply src/bitmap/util.rs:31]: I would like to add a check before the `Bound::Excluded(&i) => i - 1,` to return instead of using `saturating_sub` here.
